confmap/confmap.go

```diff
@@ -261,6 +262,36 @@ func mapKeyStringToMapKeyTextUnmarshalerHookFunc() mapstructure.DecodeHookFuncTy
 	}
 }
 
+func unmarshalerEmbeddedStructsHookFunc(_ any) mapstructure.DecodeHookFuncValue {
+	return func(from reflect.Value, to reflect.Value) (any, error) {
+		if to.Type().Kind() != reflect.Struct {
+			return from.Interface(), nil
+		}
+		fromAsMap, ok := from.Interface().(map[string]any)
+		if !ok {
+			return from.Interface(), nil
+		}
+		for i := 0; i < to.Type().NumField(); i++ {
+			if to.Type().Field(i).IsExported() && to.Type().Field(i).Anonymous {
+				if unmarshaler, ok := to.Field(i).Addr().Interface().(Unmarshaler); ok {
+					if err := unmarshaler.Unmarshal(NewFromStringMap(fromAsMap)); err != nil {
+						return nil, err
+					}
+					conf := New()
+					if err := conf.Marshal(unmarshaler); err != nil {
+						return nil, err
+					}
+					resultMap := conf.ToStringMap()
+					for k, v := range resultMap {
+						fromAsMap[k] = v
+					}
```

Comment:
Why do we need this? This is not done in `unmarshalerHookFunc`

Reply:
We run after the main unmarshaler from `unmarshalerHookFunc` and must merge with the result, rather than set it.

Reply:
1. Comment: yes,sure can do
2. oh no
3. sure can do, I believe https://github.com/open-telemetry/opentelemetry-collector/blob/main/confmap/confmap_test.go#L337 already covers this well.

2. is a big "oh no" because of #7102. We have a top level if/else that explicitly forks the flow if the struct implements Unmarshaler or not. 
If it does, it calls directly the Unmarshal function on the struct.
If it doesn't, it goes to mapstructure and then runs through the hooks.
This is this problem: https://github.com/open-telemetry/opentelemetry-collector/pull/9635#discussion_r1509814264

The workaround right now is that all structs with embedded structs that themselves implement Unmarshaler must themselves implement Unmarshaler and call mapstructure as part of their behavior.

We can straighten this up, but maybe in the next PR, if that's ok.

Reply:
Thanks for explaining and being patient with me :) Here is my proposal on what to do:

1. On this PR, we add a new (failing) test for (2) and that uses `t.Skip` with a `TODO` linking to #7102. We also add a note on #7102 about said skip to not forget about it.
2. On a future PR tackling #7102 we remove this `t.Skip`

Reply:
Thanks, added.